File: daily_problems/2025/02/0228/personal_submission/cf1267k_yawn_sean.py
# Submission link: https://codeforces.com/contest/1267/submission/308211573
def main():
    t = II()
    outs = []

    fac = [1] * 20
    for i in range(1, 20):
        fac[i] = fac[i - 1] * i

    for _ in range(t):
        k = II()
        
        cnt = [0] * 20
        length = 0
        
        for i in range(1, 21):
            cnt[k % i] += 1
            length = i
            
            k //= i
            if k == 0:
                break
        
        ans = 0
        
        to_divide = 1
        res = 1
        cur = 0
        
        for i in range(length):
            to_divide *= fac[cnt[i]]
            cur += cnt[i]
            res *= cur
            cur -= 1

        ans += res // to_divide
        
        if cnt[0]:
            cnt[0] -= 1
            length -= 1
            
            to_divide = 1
            res = 1
            cur = 0
        
            for i in range(length):
                to_divide *= fac[cnt[i]]
                cur += cnt[i]
                res *= cur
                cur -= 1
        
            ans -= res // to_divide
        
        outs.append(ans - 1)

    print('\n'.join(map(str, outs)))

File: daily_problems/2025/02/0228/personal_submission/cf1267k_yawn_sean.cpp
#include <bits/stdc++.h>
#pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/convolution"
// #include "atcoder/dsu"
// #include "atcoder/fenwicktree"
// #include "atcoder/lazysegtree"
// #include "atcoder/math"
// #include "atcoder/maxflow"
// #include "atcoder/mincostflow"
// #include "atcoder/modint"
// #include "atcoder/scc"
// #include "atcoder/segtree"
// #include "atcoder/string"
// #include "atcoder/twosat"
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;
using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's set a bit and flow!
// I came, I divide, I conquer!

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);

    vector<long long> fac(20, 1);
    for (int i = 1; i < 20; i ++)
        fac[i] = i * fac[i - 1];
    
    int t;
    cin >> t;

    while (t --) {
        long long k;
        cin >> k;

        vector<int> cnt(20);
        int length;

        for (int i = 1; i <= 20; i ++) {
            cnt[k % i] ++, length = i, k /= i;
            if (!k) break;
        }

        long long ans = 0;

        int cur = 0;
        long long to_divide = 1, res = 1;

        for (int i = 0; i < length; i ++) {
            to_divide *= fac[cnt[i]];
            cur += cnt[i];
            res *= cur;
            cur --;
        }

        ans += res / to_divide;

        if (cnt[0]) {
            cnt[0] --;
            length --;

            cur = 0, to_divide = 1, res = 1;

            for (int i = 0; i < length; i ++) {
                to_divide *= fac[cnt[i]];
                cur += cnt[i];
                res *= cur;
                cur --;
            }
    
            ans -= res / to_divide;
        }

        cout << ans - 1 << '\n';
    }

    return 0;
}